### PR TITLE
Add note about configuration to template file

### PR DESCRIPTION
**Pull request by zenizh**
For example, `config.time_zone` is referred by [ActiveSupport](https://github.com/rails/rails/blob/791bdf6fb350b5cb272e4277c1b2b3d04beb7a35/activesupport/lib/active_support/railtie.rb#L25), which initializes
before Railtie's initialize process.

They should be configured before that process, so I think Rails should note to
developers about this matter.

ref: https://github.com/rails/rails/issues/24748#issuecomment-216317145

(Please close this PR if this issue is not even a thing)

### CODE REVIEW

**Comment by prathamesh-sonpatki** [comment]
cc @fxn

File: railties/lib/rails/generators/rails/app/templates/config/application.rb [hunk 1]
```diff
@@ -25,6 +25,10 @@ class Application < Rails::Application
     # Settings in config/environments/* take precedence over those specified here.
     # Application configuration should go into files in config/initializers
     # -- all .rb files in that directory are automatically loaded.
+
+    # Some configuration such as config.time_zone should define here because
```
Comment (vpuzzella) [inline railties/lib/rails/generators/rails/app/templates/config/application.rb]: "should define" -> "should be defined"

Comment (kami-zh) [inline railties/lib/rails/generators/rails/app/templates/config/application.rb]: Thanks! 👍

**Comment by kami-zh** [comment]
Also `config.autoload_paths` which is frequently configured should be defined `config/application.rb` because of same reason.
(or add to `ActiveSupport::Dependencies.autoload_paths` directly)
ref: https://github.com/rails/rails/blob/master/railties/lib/rails/engine.rb#L565-L573

**Comment by fxn** [comment]
This is great and acknowledges an existing gotcha, but it is vague, users would wonder "OK, but which ones need to go here?".

But the problem would be that I think the answer would "all of them", and that with time the framework has deviated from this in a way I believe is not correct. There have been things moved to _config/initializers_, but in general the contract is that a railtie may check its configuration when its initializers run and configure itself based on that. Which do and which do not, that is a private implementation detail. If they have configuration endpoints, those should be ready by the time the initializers run. If you study the design, you'll see this principle is there and configuration of railties in _config/initializers_ is violating it. (I also double-checked with Yehuda and José.)

So I am not sure about what to do here, if we detail which config options need to be here, we are doing an artificial classification of railties IMO, based on a private aspect of their current implementation. The move that would be consistent with the design would be to get rid of all those _config/initializers_ and the comment that says config should go there.

Let me bring @rafaelfranca and @jeremy into the conversation, to see what they think.

**Comment by rafaelfranca** [comment]
I agree that we should change the comment saying that configurations should go in config/initializers and put the configs in application.rb. Since we changed configurations to go to config/initializers a lot of issues emerged. The main reason for those issues is that Active Record railtie and Action Controller railtie lazy load their configs but as soon a gem or application reference `ActiveRecord::Base` or `ActionController::Base` in the boot time all configs in config/initializers are ignored. It is not easy to users understand why that and they usually go in the wrong path and change the initializer file to use `ActionRecord::Base` instead of `config.active_record`.

**Comment by kami-zh** [comment]
Thank you for your comment :+1:
I was relieved to find out that my thought was not misplaced.

At least it is not good to comment that promotes to move all configurations to `config/initializers` because this confuses developers.

The best solution is to introduce a configuration that developers can understand perfectly.
But this will not be easy.

Until this is realized, I think that it is necessary to call attention to this issue on `config/application.rb`.
(I can't think of any good sentences...)

**Comment by betesh** [comment]
> The main reason for those issues is that Active Record railtie and Action Controller railtie lazy load their configs but as soon a gem or application reference ActiveRecord::Base or ActionController::Base in the boot time all configs in config/initializers are ignored.

If a component of Rails is loaded by a Railtie initializer, is there any way we could know that we're in middle of initializers?

My idea is that if Rails detects this behavior, it could output a warning with an example of the right way to do it.  Something like:


    Rails detected that ActiveRecord was loaded by initializer <INITIALIZER_NAME>.  This is
    dangerous because it will cause config.active_record settings in an application's
    config/initializers directory to be ignored.  Please wrap all references to ActiveRecord
    in your initializer in an ActiveSupport.on_load(:active_record) block.

The only question is, how would we detect that?  Is there some global state that tells us we're in middle of railtie initializers?

**Comment by betesh** [comment]
I did a little more research on this and I think that with a little work we could make it log a warning (or raise an exception if so configured) if any components are loaded before any of the initializers in Application::Finisher run.

I'll take a closer look at how much work this would take and depending on the level of effort, either open a PR and just do it, or open a new issue to discuss it first.

**Comment by matthewd** [comment]
@betesh I had wondered whether we could add a method to Configuration which would simultaneously read a supplied config key, and also lock in its value.

So if AR initializes and reads/copies settings out of the config, any later attempt to set those values in the config would give a warning/error.

**Comment by betesh** [comment]
@matthewd I think we're looking at it from opposite directions.

One option is when you're settings configuration for a Rails component, we should check whether that component has already been initialized.

The other way is, when a component is initialized, make sure it's not too early.

The advantages to the latter are that we're detecting the problem at the earliest possible point and that the responsibility of doing all this belongs to Rails initialization and not to ActiveSupport::OrderedOptions or some other type of object that's really intended to be a storage bucket.

Basically, I envision that `Rails::Application` would have some query methods that tell you its state.  Right now, all it has is `initialized?`, but we would add `initializing?` and `finishing?` (i.e. initializers from `Rails::Application::Finisher` are running and `initialized?` is still false).  If it's initializing, it's an error to load a component.  If it's finishing, then loading components is allowed.

I'm still looking into when each component is loaded to see if this will work.  Hopefully I'll have a few hours to work on it in the near future.

**Comment by matthewd** [comment]
Right.. I was just thinking that we'd simultaneously address the general issue of people trying to set/change config values too late: potentially after a "correctly timed" initialization. And we avoid explicitly tracking any "init is okay now" state -- it seems a bit closer to the root cause.

Anyway, just a thought. I look forward to seeing what you come up with. 👍🏻

**Comment by kami-zh** [comment]
I created #31572 which is related to this PR, so please check it.
